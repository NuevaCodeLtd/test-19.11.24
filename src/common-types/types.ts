import { ERROR_STATUS, Language, SUCCESS_STATUS } from '@/constant'

export type ServerSideComponentProps<Params, SearchParams = undefined> = {
  params: Params
  searchParams: SearchParams
}

export interface Post {
  id: number
  author_id: number
  title: string
  slug: string
  content: string
  summary: string
  featured_image: string
  featured_image_description: string
  featured_image_credit: string
  meta_title: string
  meta_description: string
  created_at: string // ISO 8601 string
  updated_at: string // ISO 8601 string
  tags: Tag[]
  featured: boolean
  language: Language
  relatedPosts: number[]
}

export interface Tag {
  id: number
  name: string
}

export type ServerActionResponseState =
  | {
      status: typeof SUCCESS_STATUS
      message: string
    }
  | {
      status: typeof ERROR_STATUS
      message: string
      errors?: {
        path: string
        message: string
      }[]
    }