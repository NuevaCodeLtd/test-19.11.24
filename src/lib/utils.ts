/* eslint-disable @typescript-eslint/no-explicit-any */
import path from 'path'
import fs from 'fs'
import { Post } from '@/common-types/types'

export function readFile(fileName: string) {
  const pathToFile = path.join(
    process.cwd(),
    `public/content/${fileName}`
  )
  const fileContent = fs.readFileSync(pathToFile, 'utf-8')

  return fileContent
}

export function writePostsToFile(posts: Post[]) {
  const filePath = path.join(process.cwd(), 'public/content/posts.json')
  fs.writeFileSync(filePath, JSON.stringify(posts, null, 2), 'utf-8')
}

export function formDataToObject(formData: FormData) {
  const obj: Record<string, any> = {}

  formData.forEach((value, key) => {
    // Handle multiple values
    if (obj[key]) {
      if (Array.isArray(obj[key])) {
        obj[key].push(value)
      } else {
        obj[key] = [obj[key], value]
      }
    } else {
      // Convert specific fields (e.g., checkboxes) from string to boolean
      if (value === 'true' || value === 'on') {
        obj[key] = true
      } else if (value === 'false') {
        obj[key] = false
      } else {
        obj[key] = value
      }
    }
  })

  return obj
}
