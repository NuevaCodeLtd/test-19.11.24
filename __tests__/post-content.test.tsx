import React from 'react'
import { render, screen } from '@testing-library/react'
import PostContent from '../src/components/post-item/post-content'
import { getRelatedPosts } from '@/lib/post-utils'
import { Post } from '@/common-types/types'

jest.mock('../src/lib/post-utils', () => ({
  calculateReadTime: jest.fn(),
  getRelatedPosts: jest.fn()
}))
const calculateReadTimeMock = require('../src/lib/post-utils').calculateReadTime
const getRelatedPostsMock = require('../src/lib/post-utils').getRelatedPosts
jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
  usePathname: jest.fn()
}))

describe('PostContent Component', () => {
  const mockContent = {
    id: 1,
    author_id: 101,
    title: 'Top Trends in Mobile App Development for 2024',
    slug: 'top-trends-mobile-app-development-2024',
    content:
      "<p>Mobile app development is evolving at a rapid pace, with new technologies like AI and machine learning reshaping the landscape. These trends are expected to have a significant impact on the way apps are built, deployed, and interacted with by users.</p><p>For 2024, we foresee some key trends that every developer should keep an eye on:</p><ul><li>AI integration for personalized user experiences</li><li>Progressive Web Apps (PWAs) gaining traction</li><li>5G enabling faster, more immersive apps</li></ul><p>To stay ahead of the competition, it's crucial to adopt these trends and keep your development strategies flexible.</p>",
    summary: 'Discover the top mobile app development trends for 2024.',
    featured_image: 'https://placehold.co/600x400?text=Post+1',
    featured_image_description: 'Trends in mobile development.',
    featured_image_credit: 'Photo by Unsplash',
    meta_title: 'Top Trends in Mobile App Development for 2024',
    meta_description:
      'A look at the key trends that will shape mobile app development in 2024.',
    created_at: '2024-10-21T00:00:00',
    updated_at: '2024-10-21T09:00:00',
    featured: false,
    language: 'en',
    tags: [
      { id: 1, name: 'mobile_development' },
      { id: 2, name: 'app_trends', article_id: 1 }
    ],
    relatedPosts: [2]
  } as Post
  const mockRelatedPost = {
    id: 2,
    author_id: 102,
    title: 'How AI is Revolutionizing Product Design',
    slug: 'ai-revolutionizing-product-design',
    content: '<p>Lorem ipsum</p>',
    summary:
      'Explore how AI tools are transforming the field of product design.',
    featured_image: 'https://placehold.co/600x400?text=Post+2',
    featured_image_description: 'AI in product design.',
    featured_image_credit: 'Photo by Unsplash',
    meta_title: 'How AI is Revolutionizing Product Design',
    meta_description:
      'A deep dive into the impact of AI on modern product design.',
    created_at: '2024-10-18T00:00:00',
    updated_at: '2024-10-19T09:00:00',
    featured: false,
    language: 'en',
    tags: [
      { id: 3, name: 'artificial_intelligence' },
      { id: 4, name: 'product_design', article_id: 2 }
    ],
    relatedPosts: []
  } as Post

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should match snapshot', () => {
    calculateReadTimeMock.mockReturnValue(5)
    getRelatedPostsMock.mockReturnValue(mockRelatedPost)

    const { asFragment } = render(<PostContent content={mockContent} relatedPosts={[mockRelatedPost]} />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('renders correctly with content', async () => {
    calculateReadTimeMock.mockReturnValue(5)
    getRelatedPostsMock.mockReturnValue([mockRelatedPost])

    render(<PostContent content={mockContent} relatedPosts={[mockRelatedPost]} />)

    expect(screen.getByText(mockContent.title)).toBeInTheDocument()

    expect(screen.getByText('Related Posts')).toBeInTheDocument()
    expect(
      screen.getByText('You might also be interested in:')
    ).toBeInTheDocument()
  })

  it('does not render anything if content is null', () => {
    const { container } = render(<PostContent content={null} relatedPosts={[]} />)

    expect(container).toBeEmptyDOMElement()
  })

  it('renders related posts if available', () => {
    getRelatedPostsMock.mockReturnValue(mockRelatedPost)

    render(<PostContent content={mockContent} relatedPosts={[mockRelatedPost]} />)

    expect(getRelatedPosts).toHaveBeenCalledWith(mockContent.relatedPosts)
  })

  it('does not render related posts if none are available', () => {
    getRelatedPostsMock.mockReturnValue([])

    const noRelatedContent = { ...mockContent, relatedPosts: [] }

    render(<PostContent content={noRelatedContent} relatedPosts={[]} />)

    expect(screen.queryByText('Related Posts')).not.toBeInTheDocument()
  })
})