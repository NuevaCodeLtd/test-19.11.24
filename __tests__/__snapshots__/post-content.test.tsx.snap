// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostContent Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-1tqiyaw-MuiPaper-root"
    style="--Paper-shadow: 0px 3px 1px -2px rgba(0,0,0,0.2),0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12);"
  >
    <article>
      <div
        class="MuiStack-root css-lxz2k7-MuiStack-root"
      >
        <a
          href="/"
        >
          <div
            class="MuiStack-root css-1qxw92a-MuiStack-root"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="HomeIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
                />
              </svg>
            </button>
            Go to Home Page
          </div>
        </a>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary css-v0wh71-MuiButtonBase-root-MuiButton-root"
          data-testid="delete-btn"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-1tjwu2m-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="DeleteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
              />
            </svg>
          </span>
          Delete
        </button>
        <a
          href="/edit-post/top-trends-mobile-app-development-2024"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary css-v0wh71-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-1tjwu2m-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                />
              </svg>
            </span>
            Edit
          </button>
        </a>
      </div>
      <h1
        class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1it4a80-MuiTypography-root"
      >
        Top Trends in Mobile App Development for 2024
      </h1>
      <p
        class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-2qvmir-MuiTypography-root"
      >
        Discover the top mobile app development trends for 2024.
      </p>
      <figure>
        <img
          alt="Top Trends in Mobile App Development for 2024"
          class="MuiBox-root css-qwglhd"
          src="https://placehold.co/600x400?text=Post+1"
        />
        <figcaption
          style="display: flex; align-items: center; gap: 8px;"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-1ye3urd-MuiSvgIcon-root"
            data-testid="InfoOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11 7h2v2h-2zm0 4h2v6h-2zm1-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8"
            />
          </svg>
          <span
            class="MuiTypography-root MuiTypography-caption css-1374baf-MuiTypography-root"
          >
            Trends in mobile development.  /  Photo by Unsplash
          </span>
        </figcaption>
      </figure>
      <div
        class="MuiStack-root css-82gl7o-MuiStack-root"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular css-fmeqll-MuiAvatar-root"
        >
          <img
            alt="Author 101"
            class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
            src="/images/site/avatars/avatar-5.jpg"
          />
        </div>
        <p
          class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
        >
          The article was posted: 
          <time>
            October 21, 2024
          </time>
           | 5 min read
        </p>
      </div>
      <div
        class="MuiStack-root css-18temzg-MuiStack-root"
      >
        <a
          href="/tags/mobile_development"
        >
          <div
            class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-outlinedDefault css-1d2be57-MuiChip-root"
          >
            <span
              class="MuiChip-label MuiChip-labelSmall css-19m61dl-MuiChip-label"
            >
              #mobile_development
            </span>
          </div>
        </a>
        <a
          href="/tags/app_trends"
        >
          <div
            class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-outlinedDefault css-1d2be57-MuiChip-root"
          >
            <span
              class="MuiChip-label MuiChip-labelSmall css-19m61dl-MuiChip-label"
            >
              #app_trends
            </span>
          </div>
        </a>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-zzw4lh-MuiDivider-root"
      />
      <div
        class="MuiBox-root css-4pgtvs"
      >
        <p>
          Mobile app development is evolving at a rapid pace, with new technologies like AI and machine learning reshaping the landscape. These trends are expected to have a significant impact on the way apps are built, deployed, and interacted with by users.
        </p>
        <p>
          For 2024, we foresee some key trends that every developer should keep an eye on:
        </p>
        <ul>
          <li>
            AI integration for personalized user experiences
          </li>
          <li>
            Progressive Web Apps (PWAs) gaining traction
          </li>
          <li>
            5G enabling faster, more immersive apps
          </li>
        </ul>
        <p>
          To stay ahead of the competition, it's crucial to adopt these trends and keep your development strategies flexible.
        </p>
      </div>
      <div
        class="MuiBox-root css-h5fkc8"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
        >
          Related Posts
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
        >
          You might also be interested in:
        </p>
        <ul
          class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-b9nkh6-MuiListItem-root"
          >
            <a
              href="/posts/ai-revolutionizing-product-design"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 css-o8fuj9-MuiTypography-root"
              >
                How AI is Revolutionizing Product Design
              </p>
            </a>
          </li>
        </ul>
      </div>
    </article>
  </div>
</DocumentFragment>
`;
